
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>NEAT Overview &#8212; NEAT-Python 0.92-multiparam_funcs documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.92-multiparam_funcs',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="Welcome to NEAT-Python’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="neat-overview">
<span id="neat-overview-label"></span><h1>NEAT Overview<a class="headerlink" href="#neat-overview" title="Permalink to this headline">¶</a></h1>
<p><em class="dfn">NEAT</em> (NeuroEvolution of Augmenting Topologies) is an evolutionary algorithm that creates artificial neural networks. For a
detailed description of the algorithm, you should probably go read some of <a class="reference external" href="http://www.cs.ucf.edu/~kstanley/#publications">Stanley’s papers</a> on his website.</p>
<p>Even if you just want to get the gist of the algorithm, reading at least a couple of the early NEAT papers is a good
idea.  Most of them are pretty short, and do a good job of explaining concepts (or at least pointing
you to other references that will).  <a class="reference external" href="http://nn.cs.utexas.edu/downloads/papers/stanley.cec02.pdf">The initial NEAT paper</a> is only 6 pages long, and Section II should be enough
if you just want a high-level overview.</p>
<p>In the current implementation of NEAT-Python, a population of individual <a class="reference internal" href="glossary.html#term-genome"><span class="xref std std-term">genomes</span></a> is maintained.  Each genome contains
two sets of <a class="reference internal" href="glossary.html#term-gene"><span class="xref std std-term">genes</span></a> that describe how to build an artificial neural network:</p>
<blockquote>
<div><ol class="arabic simple">
<li><a class="reference internal" href="glossary.html#term-node"><span class="xref std std-term">Node</span></a> genes, each of which specifies a single neuron.</li>
<li><a class="reference internal" href="glossary.html#term-connection"><span class="xref std std-term">Connection</span></a> genes, each of which specifies a single connection between neurons.</li>
</ol>
</div></blockquote>
<p id="index-0">To evolve a solution to a problem, the user must provide a fitness function which computes a single real number
indicating the quality of an individual genome: better ability to solve the problem means a higher score.  The algorithm
progresses through a user-specified number of generations, with each generation being produced by reproduction (either
sexual or asexual) and mutation of the most fit individuals of the previous generation.</p>
<p>The reproduction and mutation operations may add nodes and/or connections to genomes, so as the algorithm proceeds
genomes (and the neural networks they produce) may become more and more complex.  When the preset number of generations
is reached, or when at least one individual (for a <a class="reference internal" href="config_file.html#fitness-criterion-label"><span class="std std-ref">fitness criterion function</span></a> of <code class="docutils literal"><span class="pre">max</span></code>; others are configurable)
exceeds the user-specified <a class="reference internal" href="config_file.html#fitness-threshold-label"><span class="std std-ref">fitness threshold</span></a>, the algorithm terminates.</p>
<span class="target" id="index-1"></span><span class="target" id="index-2"></span><p id="index-3">One difficulty in this setup is with the implementation of <a class="reference internal" href="glossary.html#term-crossover"><span class="xref std std-term">crossover</span></a> - how does one do a crossover between two networks of differing structure?
NEAT handles this by keeping track of the origins of the nodes, with an <a class="reference internal" href="glossary.html#term-key"><span class="xref std std-term">identifying number</span></a> (new, higher numbers are generated for each additional node). Those derived from a common ancestor (that are <a class="reference internal" href="glossary.html#term-homologous"><span class="xref std std-term">homologous</span></a>) are matched up for crossover, and connections are matched if the
nodes they connect have common ancestry. (There are variations in exactly how this is done depending on the implementation of NEAT; this paragraph describes how it is done in this implementation.)</p>
<span class="target" id="index-4"></span><span class="target" id="index-5"></span><p id="index-6">Another potential difficulty is that a structural mutation - as opposed to mutations in, for instance, the <a class="reference internal" href="glossary.html#term-weight"><span class="xref std std-term">weights</span></a> of the connections - such as the addition of a node or connection can, while being promising for the future, be disruptive in the short-term (until it has been fine-tuned by less-disruptive
mutations). How NEAT deals with this is by dividing genomes into species, which have a close <a class="reference internal" href="glossary.html#term-genomic-distance"><span class="xref std std-term">genomic distance</span></a> due to similarity, then having competition most intense within species, not between species (fitness sharing). How is genomic distance measured? It uses a combination of the number of non-homologous nodes and connections with measures of how much homologous nodes and connections have diverged since their common origin. (Non-homologous nodes and connections are termed <a class="reference internal" href="glossary.html#term-disjoint"><span class="xref std std-term">disjoint</span></a> or <a class="reference internal" href="glossary.html#term-excess"><span class="xref std std-term">excess</span></a>, depending on whether the <a class="reference internal" href="glossary.html#term-key"><span class="xref std std-term">numbers</span></a> are from the same range or beyond that range; like most NEAT implementations, this one makes no distinction between the two.)</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to NEAT-Python’s documentation!</a></li>
      <li>Next: <a href="installation.html" title="next chapter">Installation</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/neat_overview.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2017, CodeReclaimers, LLC.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/neat_overview.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>